src/transformers/vocabulary2id.py

```diff
@@ -0,0 +1,91 @@
+import operator
+
+from pyspark import RDD, Row
+from models.code2vec_features import Code2VecFeatures
+
+from ast import literal_eval as make_tuple
+from sourced.ml.transformers import Transformer
+
+
+class Vocabulary2Id(Transformer):
+    def __init__(self, sc, output: str, **kwargs):
+        super().__init__(**kwargs)
+        self.output = output
+        self.sc = sc
+
+    def __call__(self, rows: RDD):
+        value2index_freq, path2index_freq = self.build_vocabularies(rows)
+
+        doc2path_contexts = self.build_doc2pc(value2index_freq, path2index_freq, rows)
+
+        doc2path_contexts = doc2path_contexts.collect()
+
+        Code2VecFeatures().construct(value2index_freq=value2index_freq,
+                                     path2index_freq=path2index_freq,
+                                     path_contexts=doc2path_contexts).save(
+            self.output)
+
+    @staticmethod
+    def _unstringify_path_context(row):
+        """
+        Takes a row containing ((pc, doc), freq) and returns a tuple (u, path, v)
+        (removes namespace prefix v.)
+        """
+        return make_tuple(row[0][0][2:])
+
+    def build_vocabularies(self, rows: RDD):
+        """
+        Process rows to gather values and paths with their frequencies.
+        :param rows: row structure is ((key, doc), val) where:
+            * key: str with the path context
+            * doc: file name
+            * val: number of occurrences of key in doc
+        """
+
+        def _flatten_row(row: Row):
+            # 2: removes the namespace v. from the string to parse it as tuple
+            k = Vocabulary2Id._unstringify_path_context(row)
+            return [(k[0], 1), (k[1], 1), (k[2], 1)]
+
+        rows = rows \
+            .flatMap(_flatten_row) \
+            .reduceByKey(operator.add) \
+            .persist()
+
+        values = rows.filter(lambda x: type(x[0]) == str).collect()
+        paths = rows.filter(lambda x: type(x[0]) == tuple).collect()
+
+        value2index_freq = {w: (i, count) for i, (w, count) in enumerate(values)}
+        path2index_freq = {w: (i, count) for i, (w, count) in enumerate(paths)}
+
+        rows.unpersist()
+
+        return value2index_freq, path2index_freq
+
+    def build_doc2pc(self, value2index_freq: dict, path2index_freq: dict, rows: RDD):
+        """
+        Process rows and build elements (doc, [path_context_1, path_context_2, ...])
+        :param value2index_freq: value -> (id, freq)
+        :param path2index_freq: path -> (id, freq)
+        """
+
```

Comment:
you are broadcasting frequencies uselessly here, better to use dict comprehension beforehand:
```
value2index = self.sc.broadcast({key: id for key, (id, _) in value2index_freq.items()})
```